component/file_cache/file_cache.go

```diff
@@ -1268,11 +1374,25 @@ func (fc *FileCache) TruncateFile(options internal.TruncateFileOptions) error {
 	} else {
 		// If size is not 0 then we need to open the file and then truncate it
 		// Open will force download if file was not present in local system
-		h, err = fc.OpenFile(internal.OpenFileOptions{Name: options.Name, Flags: os.O_RDWR, Mode: fc.defaultPermission})
+
+		h, err = fc.OpenFile(internal.OpenFileOptions{Name: options.Name, Flags: 0, Mode: fc.defaultPermission})
 		if err != nil {
-			log.Err("FileCache::TruncateFile : Error opening file %s [%s]", options.Name, err.Error())
-			return err
+			log.Err("FileCache::TruncateFile : Error calling OpenFile with %s [%s]", options.Name, err.Error())
 		}
+		if _, loaded := h.GetValue("flag"); loaded {
+			_, err = fc.DownloadFile(internal.DownloadFileOptions{Name: options.Name, Handle: h})
+			if err != nil {
+				log.Err("FileCache::TruncateFile : Error opening file %s [%s]", options.Name, err.Error())
+				return err
+			}
+		} else {
+			h, err = fc.OpenFile(internal.OpenFileOptions{Name: options.Name, Flags: os.O_RDWR, Mode: fc.defaultPermission})
```

Comment:
when the file gets flushed, it needs write permissions. I just updated the initial OpenFile() call to also include RDWR flag before it gets downloaded. the flags are now consistent. Does this seem correct to you?

Reply:
Yes, but I don't see why you need an open call on 1378 and then another open call on line 1389.

Reply:
the first `OpenFile()` call is for the `DownloadFile()` to take place. This came from Michael's suggestion to call Open instead of creating a new Handle directly like I was doing previously. 

The second `OpenFile()` call might not be necessary since it was already called previously. That was there previously. The entire else block might not be needed anymore. Do you agree?

Reply:
After looking at this again, I think you can simplify it further. Here I think you are trying to think about it too hard. So you are determining here is the file needs to be downloaded, but there is a ton of logic built into DownloadFile already to figure that out for you. So don't worry about it here. Just call the first OpenFile and then DownloadFile after. It will download it if needed, and it will skip it if not.

Reply:
I might suggest running this by Michael just to flesh out the design a bit, but my thinking is that you should either have DownloadFile be responsible for checking if a download is required or not (probably prefereable) or the caller of DownloadFile is responsible for checking if a download is required and then it will download it. Right now you are doing a bit of both.